lib/client.js

```diff
@@ -854,6 +858,12 @@ class Parser {
       } else {
         client[kKeepAliveTimeoutValue] = client[kKeepAliveDefaultTimeout]
       }
+
+      const max = this.keepAlive ? util.parseKeepAliveMax(this.keepAlive) : null
+
+      if (max != null) {
+        client[kPipelining] = Math.min(client[kMaxPipelining], max)
```

Comment:
I think we should be significantly more conservative than this, as the closer we are to the maximum number of request left the higher the risk of going over the limit. Something like the following would do:

```suggestion
        client[kPipelining] = Math.min(client[kMaxPipelining], max >= client[kMaxPipelining] ? max >= client[kMaxPipelining] : 1)
```